tests/late_joiners.py

```diff
@@ -104,107 +107,106 @@ def run_requests(nodes, total_requests, start_id, final_msg, final_msg_id):
                 node_id += 1
                 c = clients[node_id % len(clients)]
                 try:
-                    c.rpc("LOG_record", {"id": id, "msg": long_msg})
+                    check_commit(
+                        c.rpc("LOG_record", {"id": id, "msg": long_msg}), result=True
+                    )
                 except (TimeoutError, requests.exceptions.ReadTimeout,) as e:
-                    LOG.info("Trying to access a suspended node")
+                    LOG.info("Trying to access a suspended network")
+                    if cant_fail:
+                        raise RuntimeError(e)
                 id += 1
+
         wait_for_nodes(nodes, final_msg, final_msg_id)
 
 
 def run(args):
-    hosts = ["localhost", "localhost", "localhost"]
+    hosts = ["localhost", "localhost", "localhost", "localhost"]
 
     with infra.ccf.network(
         hosts, args.binary_dir, args.debug_nodes, args.perf_nodes, pdb=args.pdb
     ) as network:
         network.start_and_join(args)
         first_node, backups = network.find_nodes()
         all_nodes = network.get_joined_nodes()
-
         term_info = find_primary(network)
         first_msg = "Hello, world!"
         second_msg = "Hello, world hello!"
+        catchup_msg = "Hey world!"
         final_msg = "Goodbye, world!"
 
-        nodes_to_kill = [backups[0]]
-        nodes_to_keep = [first_node, backups[1]]
-
-        # first timer determines after how many seconds each node will be suspended
-        if not args.skip_suspension:
-            timeouts = []
-            for node in all_nodes:
-                t = random.uniform(1, 10)
-                LOG.info(f"Initial timer for node {node.node_id} is {t} seconds...")
-                timeouts.append((t, node))
-
-            for t, node in timeouts:
-                tm = Timer(t, timeout, args=[node, True, args.election_timeout / 1000],)
-                tm.start()
-
-        LOG.info(
-            "Adding another node after f = 0 but before we need to send append entries"
-        )
-        # check that a new node can catch up naturally
-        new_node = network.create_and_trust_node(
-            lib_name=args.package, host="localhost", args=args,
-        )
-        assert new_node
-        nodes_to_keep.append(new_node)
-
         with first_node.node_client() as mc:
             check_commit = infra.checker.Checker(mc)
             check = infra.checker.Checker()
 
-            run_requests(all_nodes, TOTAL_REQUESTS, 0, first_msg, 1000)
+            run_requests(all_nodes, int(TOTAL_REQUESTS / 2), 0, first_msg, 1000)
             term_info.update(find_primary(network))
 
-            # check that new node has caught up ok
-            assert_node_up_to_date(check, new_node, first_msg, 1000)
-            # add new node to backups list
-            all_nodes.append(new_node)
+            nodes_to_kill = []
+            nodes_to_kill.append(backups[0])
+            nodes_to_keep = [first_node, backups[1], backups[2]]
```

Comment:
Since we move to our test suite and chainable tests, it is better to have each test unaware of the number of nodes that the `network` it is given has (the only requirement should be a minimum number of nodes, `@at_least_n_nodes(n)`).

```
node_to_kill = network.find_any_backup()
node_to_keep = [n for n in network.get_joined_nodes() if n != node_to_kill]
```

Reply:
Sure thanks :)

Reply:
This actually works pretty nicely since the test won't kill the same node each time :)